### PR TITLE
Fix compiler: use reference checks for equality

### PR DESCRIPTION
**Pull request by tbosch**
Supersedes #11111

### CODE REVIEW

**Comment by tbosch** [comment]
First commit is LGTMed by me. Only second commit needs LGTM from somebody else.

File: modules/@angular/compiler/src/compile_metadata.ts [hunk 1]
```diff
@@ -682,20 +571,20 @@ export class TransitiveCompileNgModuleMetadata {
...
-    pipes.forEach(pipe => this.pipesSet.add(pipe.type.runtime));
+    directives.forEach(dir => this.directivesSet.add(dir.type.reference));
+    pipes.forEach(pipe => this.pipesSet.add(pipe.type.reference));
   }
 }
 
 export function removeIdentifierDuplicates<T extends CompileMetadataWithIdentifier>(items: T[]):
     T[] {
-  const map = new CompileIdentifierMap<T, T>();
+  const map = new Map<T, T>();
```
Comment (chuckjaz) [inline modules/@angular/compiler/src/compile_metadata.ts]: Consider using Set<T> instead.

**Comment by chuckjaz** [comment]
Minor nit, otherwise LGTM

File: modules/@angular/compiler/src/compile_metadata.ts (hunk 1 above)
Comment (tbosch) [inline modules/@angular/compiler/src/compile_metadata.ts]: Thanks

**Pull request closed without merging**